### ISSUE TITLE
Multiple compiler support?

### ISSUE THREAD

**Issue by SegaraRai** [antfu/unplugin-icons#253]
### Clear and concise description of the problem

Especially in a meta-framework like Astro, where we can mix frameworks, we may want to import icons as Astro components in the .astro file and as Vue components in the .vue file.
Currently we can retrieve raw SVG text using `?raw` query, but I think it would be nice to have a more flexible choice of formats.

### Suggested solution

If an extension is specified, use the appropriate compiler for that extension.
This would be a breaking change, so we need to make this feature opt-in.

examples:

```ts
Icons({
  useExtension: false, // default
})

Icons({
  useExtension: true,
})

Icons({
  useExtension: {
    // custom overrides
    '.vue': 'vue2',
  },
})
```

### Alternative

Select compiler by `?compiler=astro` query.
In this solution, we probably do not need to make any changes to the `options`.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/antfu/.github/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://github.com/antfu/contribute).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by jwing8** [comment]
I was able to use icons in both .astro and .vue files by configuring the Icons compiler property (under vite.plugins) to 'vue3'.  I was also able set the IconsResolver (under vite.plugins.Components.resolvers) to allow for icon pack aliases and a custom template prefix to use.  These settings are working for both Astro and Vue components.  As a bonus, the 'unplugin-vue-components' package allows for auto-importing icons within the Vue components (still have to manually import in the Astro files though).

This addressed my use case but I don't think this approach would work should you have multiple UI integrations (e.g. Vue + Svelte), so something like your proposed solution would make sense.

**Comment by DeadOce4n** [comment]
I'm using this in a project which has both astro and svelte components, it works correctly without extra configuration by simply calling the Icons function twice inside the plugins array, one for svelte and one for astro, but the one for svelte should be first:
```ts
export default defineConfig({
  integrations: [svelte()],
  vite: {
    plugins: [
      Icons({
        compiler: 'svelte',
        autoInstall: true,
      }),
      Icons({
        compiler: 'astro',
        autoInstall: true,
      }),
    ],
  },
});

```

**Comment by Trombach** [comment]
> I'm using this in a project which has both astro and svelte components, it works correctly without extra configuration by simply calling the Icons function twice inside the plugins array, one for svelte and one for astro, but the one for svelte should be first:
> 
> ```ts
> export default defineConfig({
>   integrations: [svelte()],
>   vite: {
>     plugins: [
>       Icons({
>         compiler: 'svelte',
>         autoInstall: true,
>       }),
>       Icons({
>         compiler: 'astro',
>         autoInstall: true,
>       }),
>     ],
>   },
> });
> ```

@DeadOce4n This does seem to work, thanks! Do you use typescript? I can't figure out how to set up type declarations for this. The code compiles fine, but I get typescript errors in either `.astro` or `.svelte` files. If I add
```json
// tsconfig.json
{ 
  "compilerOptions": {
    "types": [
      "unplugin-icons/types/astro",
    ]
  }
}
```
as per the docs, my astro imports show no errors. But as soon as I try to add the svelte types the astro types break again. I've tried adding `"unplugin-icons/types/svelte"` to the `"types"` array in `tsconfig.json` as well as adding `import 'unplugin-icons/types/svelte'` or `/// <reference types="unplugin-icons/types/svelte" />`, as well as other things, but nothing works.

Any ideas how to fix this?

**Comment by Visual-Dawg** [comment]
Subscribing here too for the Typescript solution

**Comment by Visual-Dawg** [comment]
I solved it by using `~icons/*` for `Astro` and `virtual:icons/*` for `Svelte`.

To achieve that I put the needed contents of the referenced `.d.ts` files directly into `env.d.ts` like this:

```ts
/// <reference path="../.astro/types.d.ts" />
/// <reference types="astro/client" />
/// <reference types="astro/astro-jsx" />

declare module "virtual:icons/*" {
  import { SvelteComponent } from "svelte"
  import type { SvelteHTMLElements } from "svelte/elements"

  export default class extends SvelteComponent<SvelteHTMLElements["svg"]> {}
}

declare module "~icons/*" {
  const component: (props: astroHTML.JSX.SVGAttributes) => astroHTML.JSX.Element
  export default component
}
```

**Comment by arthurrmp** [comment]
> I solved it by using `~icons/*` for `Astro` and `virtual:icons/*` for `Svelte`.

Thanks for that @Visual-Dawg! I've been able to use icons for React and Astro using that tip.

If anyone is wondering, first, if you have imported the types on your tsconfig.json, remove it:
```json
{
  "extends": "astro/tsconfigs/strict",
  "compilerOptions": {
    "baseUrl": "/Users/arthur/Documents/Projetos/Khada/khada-astro",
    "paths": {
      "@/*": [
        "./src/*"
      ]
    },
    "jsx": "react-jsx",
    "jsxImportSource": "react",
-    "types": [
-      "unplugin-icons/types/astro",
-   ]
  }
}
```

Then add this at the end of your  env.d.ts file:
```typescript
//...

declare module "virtual:icons/*" {
  import type { SVGProps } from "react";
 import type React from "react";

  const component: (props: SVGProps<SVGSVGElement>) => React.ReactElement;
  export default component;
}

declare module "~icons/*" {
  const component: (
    props: astroHTML.JSX.SVGAttributes,
  ) => astroHTML.JSX.Element;
  export default component;
}

```

To use an icon from a .tsx file:
```tsx
import MdiAlarmOff from "virtual:icons/mdi/alarm-off";

export default function Test() {
  return (
    <MdiAlarmOff />
  )
}
```

To use an icon from a .astro file:
```astro
---
import CarbonSun from "~icons/carbon/sun";
---

<CarbonSun />

```

**Comment by Trombach** [comment]
Thanks @Visual-Dawg and @DeadOce4n for your solutions! I have gone with the vite resolve solution for now.

@antfu I'm wondering if this could maybe be improved either on the types or at least the documentation side? Since one of Astro's main strengths is that you can use other framework components within it, I think it would be great if the astro types that are shipped with this library allow you to import icons into all supported framework components.

I believe, one solution would be to add `/icons:svelte/`, `/icons:astro/` etc to the allowed `URL_PREFIXES` array [here](https://github.com/unplugin/unplugin-icons/blob/224e1e5ff47968259bdfecded64a850fde2b705b/src/core/loader.ts#L7) and then declare all framework specific modules in the astro type file.
There are probably better solutions and I'm open for suggestions. I'm happy to provide a pull request for this if I get some guidance on how this should be handled.

**Comment by MrSquaare** [comment]
Did someone manage to use it with Astro and React? Because it doesn't work since unplugin is trying to use an Astro component instead of a React component

```
Astro components cannot be rendered directly via function call, such as Component() or {items.map(Component)}.
```

**Comment by rishi-raj-jain** [comment]
> Did someone manage to use it with Astro and React? Because it doesn't work since unplugin is trying to use an Astro component instead of a React component
> 
> ```
> Astro components cannot be rendered directly via function call, such as Component() or {items.map(Component)}.
> ```
> 
> EDIT: Even mixing React and Vue doesn't work, so indeed we really need a way to tells Unplugin which compiler it should use (via import query or path matching)

Yes! I wrote a quick [blog](https://www.launchfa.st/blog/react-astro-unplugin-icons) on it.

**Comment by adamhl8** [comment]
I've been trying to find a way to use icons in both Astro and React components. I tried what DeadOce4n suggested here https://github.com/unplugin/unplugin-icons/issues/253#issuecomment-1918126322, but trying to use the `class` prop on icons didn't work in Astro components.

After some messing around, this is the simplest way I found to get icons working in both:

`astro.config.*`
```ts
import react from "@astrojs/react"
import { defineConfig } from "astro/config"
import Icons from "unplugin-icons/vite"

export default defineConfig({
  vite: {
    // https://github.com/unplugin/unplugin-icons/issues/253#issuecomment-1918126322
    plugins: [Icons({ compiler: "jsx", jsx: "react" })], // enable `autoInstall` if you want
  },
  integrations: [react()],
})
```

`icons.d.ts` (doesn't have to be named `icons`)
```ts
/// <reference types="unplugin-icons/types/react" />
```

With this, all icons are of course React components (even in Astro components, so use `className`), and this only works if you are using the `react` Astro integration.

**Comment by tisonkun** [comment]
@adamhl8 Thanks for your solution! Would it require to use `client:load="react"` when using the Icons components in an astro component?

**Comment by adamhl8** [comment]
> Thanks for your solution! Would it require to use `client:load="react"` when using the Icons components in an astro component?

Nope, the icons shouldn't need any `client:*` directives since they're not interactive (i.e. they don't use any JS).

**Comment by gustadev24** [comment]
Any update? I can confirm the problem that it was explained by @adamhl8 in [#253 comment](https://github.com/unplugin/unplugin-icons/issues/253#issuecomment-2993807766)
